### PR TITLE
Bugfix/issue 2772 mcmc multiunit compilation

### PR DESCRIPTION
**Pull request by fatvlady**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Fixes issue 2772 about multi-unit compilation of base_xhmc.hpp.
Made `stable_sum` return resulting pair instead of writing to out `double&` parameters. Utilizing C++17 structured bindings would make it better.
Moved `stable_sum` inside base_xhmc scope implicitly making it `inline`. Consider to move to math module, if can be reused?

#### Intended Effect
No linker duplicate symbols errors when base_xhmc.hpp is included in several translation units.

#### How to Verify
`./runTests.py src/test/unit/multiple_translation_units_test`

Makefile was adjusted to allow different multi-unit compilation tests. To add new multi-unit test, the one has to specify new `libmtu_*.so` dependency, and add corresponding `mtu_*.cpp` test to `src/test/unit/`, Make will automatically create 2 copies of `mtu_*.cpp`, compile them and link into one shared object.

#### Documentation
Added [in] modifier to stable_sum function parameters.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Vlad Fatenko, under BSD 3-clause; documentation is inline, under CC-BY 4.0.

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Thanks for submitting a PR.  We really appreciate it and should be able to get this merged.

This PR does two things:

1. resolves the multi-unit compilation issue,
2. refactors `stable_sum` to use `std::tie`

I'd be OK reviewing (1), but (2) I think we need @betanalpha.  

Also, before we can merge, we need a declaration of who the copyright holder is and that they are releasing the code/doc under BSD/CC-BY.  Sorry to be so bureaucratic about this, but we're trying very hard to keep our copyright trail clean.

**Comment by fatvlady** [comment]
Added my name to copyright section in summary.

**Comment by bob-carpenter** [comment]
Thanks for the copyright notice.  This one's ready to review.  It looks OK to me, but touches things in the algorithms and makefiles about which I'm less confident.

@betanalpha---can you take a look at the use of `std::tie` and see if it's OK?

@syclik --- can you take a look at the makefiles?

**Comment by betanalpha** [comment]
Code looks good to me.

Some strong style recommendations in base_xhmc.hpp:

Line 106: align “ave_subtree” with “ave” in stable_sum on line 105

Lines 166-169: add additional space to align arguments (because bool is one character longer than int)

Line 186: align “dG_dt” with “ave” in stable_sum on line 185

Line 211: align “ave_subtree” with “ave” in stable_sum on line 210

Line 226: align “ave_subtree” with “ave” in stable_sum on line 225

Line 231: change to

std::tie(ave_subtree, log_sum_weight_subtree) 
  = stable_sum(ave_left, log_sum_weight_left, 
                         ave_right, log_sum_weight_right);

Line 265: add space before first “double”

Line 269: align “log_w2” with “(e” on line 268

Line 273: align “log_w1" with “(a1” on line 268

**Comment by bob-carpenter** [comment]
Could you make that an official review through GitHub?  We need at least one reviewer for every PR before merging.

**Comment by fatvlady** [comment]
@betanalpha There is an option in clang-format, called `AlignAfterOpenBracket`. Does exactly what you asked automatically when set to `BAS_Align`. Might be useful to incorporate along with `cpplint`.

**Comment by bob-carpenter** [comment]
> There is an option in clang-format, called AlignAfterOpenBracket. Does exactly what you asked automatically when set to BAS_Align

Yes, please.  I'm not sure how we'd stage this as it'd presumably rewrite a bunch of our existing code.

**Comment by fatvlady** [comment]
Any updates on this PR? Seems like Jenkins is broken.

**Comment by betanalpha** [comment]
We are currently holding off on enabling clang-format on the non-math
libraries until we get the monorepo going so for now we’ll have to 
handle the formatting by hand.

I am currently slammed with other Stan obligations but will do a full
review that covers the new testing as well next week.

File: src/stan/mcmc/hmc/xhmc/base_xhmc.hpp [hunk 1]
```diff
@@ -283,7 +284,7 @@ namespace stan {
       double x_delta_;
 
       int n_leapfrog_;
-      int divergent_;
+      bool divergent_;
```
Comment (betanalpha) [inline src/stan/mcmc/hmc/xhmc/base_xhmc.hpp]: For future reference we recommend PRs isolate one fix at a time and avoid aggregating small, unrelated updates like this.

Comment (fatvlady) [reply src/stan/mcmc/hmc/xhmc/base_xhmc.hpp]: Ok, just fixed problems discussed in related issue page.

**Comment by fatvlady** [comment]
Could you please restart Jenkins?

**Comment by betanalpha** [comment]
The review still isn’t complete — I’m following up
on something regarding the MTU test.  Once the
formal review is submitted I’ll make sure the CI
gets going.

**Comment by betanalpha** [comment]
Just following up — the review is in and Jenkins should be
fixed and automatically rerun once the small changes are 
made and pushed (but I’ll watching over it to confirm).

**Comment by fatvlady** [comment]
Ok, I will move tests today or tomorrow.

**Comment by fatvlady** [comment]
Seems like Jenkins is hung.

**Comment by betanalpha** [comment]
Right now the cloud Windows instances aren’t working so
we’re limited to a single local Windows machine for all of
the Windows testing.  That machine has been flakey in the
past, as seems to have flaked out for this test, but also we
have a pretty long queue for testing through it at the moment.

I’ll keep an eye on it to usher it through.

**Comment by betanalpha** [comment]
The integration tests aren't passing -- once you moved them did you verify that runTests.py src/test/interagration worked?

**Comment by fatvlady** [comment]
Pushed typo... Fixed now.

**Comment by betanalpha** [comment]
Thanks!

**Comment by fatvlady** [comment]
Thanks for maintaining this library!

**Pull request merged**